### ISSUE TITLE
Timed out waiting for handshake

### ISSUE THREAD

**Issue by eagress** [theophilusx/ssh2-sftp-client#266]
Debug doesn't give me much information. The code connects and then times out waiting for handshake. 

Is there any chance you could give me some advice on what to try? I don't know what's going on in IN_INIT.

I've attached the node code as a txt file. It's only trying to connect to the server. Nothing fancy.
The debug log is in testit debug.txt.

[testit.txt](https://github.com/theophilusx/ssh2-sftp-client/files/5264288/testit.txt)
[testit debug.txt](https://github.com/theophilusx/ssh2-sftp-client/files/5264296/testit.debug.txt)

I've tried commenting out the algorithms and let the code use its defaults. That made no difference.

Thanks.

**Comment by theophilusx** [comment]
Version of node?
Version of ssh2-sftp-client?
Platform client is on?
Platform remote sftp server is on?

from that debug I'd suspect your not even getting close to connecting to
the host. There is typically a lot more backward and forward negotiation
before authentication and encryption schemes are decided upon.

 I would remove the algorithms property from the connect object. I don't
 think that is the issue as I don't think your even getting that far.

First thing I would do is use openSSH's sftp CLI and try connecting
using the exact same credentials and host address, but include the -v
(verbose) flag. This will verify there are no other issues, such as
firewalls. The verbose flag will also tell you a bit about the remote
sftp server, including what algorithm, key exchange etc it supports,
just in case you do need to set these values (though you will need to
make sure you map to the names the ssh2 module uses). It may be
necessary to give more than one 'v' to increase verbosity. 

If you are able to connect with sftp, then I would try connecting using
one of the scripts in the tool directory of the repo (e.g.
lowlevel-stat.js or lowlevel-realpath.js) which only use the ssh2 module
(no ssh2-sftp-client). If that still fails, then we know the issue is
at the ssh2 level, if it works, then it is at the ssh2-sftp-client level
and we can investigate further.

**Comment by eagress** [comment]
Version of Node - 13.9.0
Version of ssh2-sftp-client - 5.3.1
Client platform is Windows 10 Pro
I don't know the specifics of the server's platform. I'm looking for a technical contact that might be able to provide some details.

My very neophyte attempt to test with ssh on Windows is attached.

A password is required but I never discovered a way to provide it during the connection attempt. After 5 minutes the server times out and the connect closes.

Thank you for looking at this. Since the first time you helped me on another issue, things have gone very well. I'm hoping you will see something that gets me past this issue too.

[Windows ssh attempt 1.txt](https://github.com/theophilusx/ssh2-sftp-client/files/5271332/Windows.ssh.attempt.1.txt)

**Comment by theophilusx** [comment]
From that debug output, my guess is that you are not allowed to ssh into
that server on that port. It also could be that your only allowed to
connect if you use either an ssh key or an ssh certificate. sometimes,
if admins don't want you allow ssh but do want to allow sftp, they will
configure things such that you can only connect to ssh if you provide an
ssh key while allowing sftp with a password. 

Try using sftp rather than ssh and see what result you get. You should
still be able to use the -v switch for debug info.

The port your trying to connect to is also a non-standard port. Is there
a reason you have to use port 990? Were you given any specific
instructions/details on connecting to this server?

Both ssh and sftp are very powerful protocols with many options. At this
stage, my suspicion is your missing something required in order to
connect to this server or there is some kind of network issue (e.g.
firewall). My recommendation would be to get things working with
openSSH's sftp and once you have that ironed out, go back to your node
script and get it working there.

If you cannot get sftp to work immediately, read up on ssh and sftp to
become more familiar with the protocols and what is supported (for
example, using ssh keys, ssh certificates, understanding some of the
configuration options, like strictMode etc. Most people only know
about/understand about 10% of ssh/sftp, which is fine most of the time.
However, once you begin to run into issues, you need to dig deeper. This
can hold for configuration of the server end as well. For example, you
can setup both ssh and sftp so that it will only work with a key and/or
a certificate, so check any instructions provided by the admins to be
sure you haven't missed something. If you cannot get things to work with
the openSSH sftp CLI, contact the server admins and ask them for help.
However, make sure you have read up and understood (or know what you
don't understand) ssh and sftp before you contact them. They will likely
assume you understand things like they do and if you can't ask the right
questions, you may not get the right answers or the answers you get
won't make sense.

**Comment by eagress** [comment]
From what I've been able to learn, the server requires FTP over SSL. The SSL connection has to be establish first before any command is sent to the server. I get the same result when I try ssh or sftp CLI. The last message from the server is 550 SSL/TLS required on the control channel. After that nothing more happens until the login times out. I think the timeout might actually be on my side and perhaps not server initiated.

It's like the server is sitting there waiting for something I'm not sending in order to establish the encrypted connection before it even allows a handshake. Does that even make sense?

**Comment by theophilusx** [comment]
It is important to recognise that sftp is not the same as FTP over SSL.
When FTP was first introduced, it was like telnet and did not use any
form of encryption. Later when it was realised encryption was required,
there were 2 approaches commonly used.

The first approach was to 'tunnel' the existing protocol (i.e. telnet
and ftp) over an SSL connection (later TLS). This was often referred to
as telnets and ftps.

The other approach was to use new protocols which had security built in
(i.e. ssh to replace telnet and sftp to replace ftp).

In the first approach, first an SSL or TLS tunnerl between the two ends
is established and once it is in place, the 'normal' telnet or ftp
protocol runs over that tunnel.

In the second, the encrypted channel is established as part of the
initial connection - it is all part of the one protocol.

The first approach, tunnelling ftp over SSL/TLS is less common these
days, but does still exist. However, ftps is not compatible with sftp.

The SFTP protocol (based on ssh) supports multiple different
authentication methods, including passwords, ssh keys and certificates.

If the remote server your trying to connect to is ftps and not sftp, you
won't be able to use ssh2-sftp-client or openSSH sftp CLI. If on the other hand, the remote
server is sftp, but is configured to use a certificate, you will need to
configure the sftp client to use the certificate. You would also need to
be given the certificate by the server admins or there would need to be
some URL where you can obtain the certificate.

You need to verify whether the remote server is ftps or sftp. If it is
sftp, but requires a certificate, you need to obtain the certificate and
configure openSSH sftp CLI to use that certificate (command line switch
I think). Once you have that working, then you can work out how to
configure ssh2-sftp-client to connect using a certificate (I believe
ssh2/ssh2-streams does support this, but have never used it).

If on the other hand, the remote server turns out to be ftps and not
sftp, you will need to find a different node library/module. It could be
done using a node TLS module and a plain ftp module (theoretically, I've
never done it, but the process is fairly straigh-forward).

**Comment by theophilusx** [comment]
Is there anything else required here or can this issue be closed?

**Comment by eagress** [comment]
Yes, you can close this. It's FTPS. I finally wrote something in C# to figure it out. Sorry I didn't update you. As always, I appreciate your help.

Sent from my Verizon, Samsung Galaxy smartphone
Get Outlook for Android<https://aka.ms/ghei36>

________________________________
